Kubernetes source plugin

huseyinbabal opened the pull request:
<!-- Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines, templates, the recommended Git workflow, and any related documentation.
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->
⚠️ Kubernetes plugin extraction was a huge task, but we couldn't figured it out in the beginning. So, you see a failing tests (2 of them only), but I intentionally left them as is since I have a couple of followup tasks for this PR. You can see them all [here](https://github.com/kubeshop/botkube/issues/840)  

## Description

Changes proposed in this pull request:

- Kubernetes plugin is implemented
- Filters are removed from core configuration and  moved to kubernetes plugin
- Actions are managed in core, not in kubernetes source anymore
- Recommendations are part of kubernetes plugin, they are removed from core
- Please see followup tasks [here](https://github.com/kubeshop/botkube/issues/840) 

## Testing

<!-- Describe necessary steps to test the changes.
You can refer to the existing documentation if some steps are already described. -->

### Build Kubernetes Plugin
Following command will build kubernetes plugin 
```shell
PLUGIN_DOWNLOAD_URL_BASE_PATH=http://localhost:8080/plugin-dist make gen-plugins-index
```

### Run Plugin Server
```shell
npx serve --listen 8080
```

### Run Botkube With the Following Configurations
```yaml

communications:
  'default-group':
    # Settings for Slack
     socketSlack:
       enabled: true
       appToken: "xapp-..."
       botToken: "xoxb-..."
      notification:
        type: "long" # TODO: Change to `short` during testing
      channels:
        'default':
          name: botkube-demo
          bindings:
            sources:
              # - plugin-based
              # - k8s-events
              - k8s-annotated-cm-delete
              - k8s-updates
              # - k8s-pod-create-events
                #              - plugin-based
            executors:
              - kubectl-plugin
        'secondary':
          name: botkube-demo-secondary
          bindings:
            sources:
              # - plugin-based
              - k8s-events
              # - k8s-annotated-cm-delete
              # - k8s-pod-create-events
                #              - plugin-based
            executors:
              - kubectl-plugin

settings:
  clusterName: gke-playground
  configWatcher: true
  lifecycleServer:
    deployment:
      name: botkube
      namespace: botkube
    port: "2113"

configWatcher:
  enabled: false
  initialSyncTimeout: 0

sources:
  'k8s-events':
    displayName: "K8s recommendations"
    'botkube/kubernetes':
      enabled: true
      config:
        kubeConfig: /Users/huseyin/.kube/config
        log:
          level: debug
        recommendations:
          pod:
            noLatestImageTag: true
            labelsSet: true
          ingress:
            backendServiceValid: false
            tlsSecretValid: false
        namespaces:
          include:
            - botkube
        event:
          types:
            - create
            - update
        resources:
          - type: v1/configmaps
  'k8s-annotated-cm-delete':
    displayName: "K8s ConfigMap delete events"
    'botkube/kubernetes':
      enabled: true
      config:
        kubeConfig: /Users/huseyin/.kube/config
        log:
          level: debug
        namespaces:
          include:
            - botkube
        # labels:
        #   test.botkube.io: "true"
        event:
          types:
            - delete
        resources:
          - type: v1/configmaps

  'k8s-pod-create-events':
    'botkube/kubernetes':
      enabled: true
      config:
        log:
          level: debug
        kubeConfig: /Users/huseyin/.kube/config
        namespaces:
          include:
            - botkube
        event:
          types:
            - create
        resources:
          - type: v1/pods

  'k8s-updates':
    displayName: "K8s ConfigMaps updates"
    'botkube/kubernetes':
      enabled: true
      config:
        log:
          level: debug
        kubeConfig: /Users/huseyin/.kube/config
        namespaces:
          include:
            - default
        event:
          types:
            - create
            - update
            - delete
        resources:
          - type: v1/configmaps
            namespaces:
              include:
                - botkube
            event: # overrides top level `event` entry
              types:
                - update
            updateSetting:
              includeDiff: true
              fields:
                - data
executors:
  'kubectl-plugin':
    ## Kubectl executor configuration.
    botkube/kubectl:
      enabled: true
      # namespaces:
      #   # -- List of allowed Kubernetes Namespaces for command execution.
      #   # It can also contain a regex expressions:
      #   #  `- ".*"` - to specify all Namespaces.
      #   include:
      #     - ".*"
      #   # -- List of ignored Kubernetes Namespace.
      #   # It can also contain a regex expressions:
      #   #  `- "test-.*"` - to specify all Namespaces.
      #   exclude: []
      # # -- If true, enables `kubectl` commands execution.
      # enabled: true
      # ## List of allowed `kubectl` commands.
      # commands:
      #   # -- Configures which `kubectl` methods are allowed.
      #   verbs: ["api-resources", "api-versions", "cluster-info", "describe", "explain", "get", "logs", "top"]
      #   # -- Configures which K8s resource are allowed.
      #   resources: ["deployments", "pods", "namespaces", "daemonsets", "statefulsets", "storageclasses", "nodes", "configmaps", "services", "ingresses"]
      # # -- Configures the default Namespace for executing Botkube `kubectl` commands. If not set, uses the 'default'.
      # defaultNamespace: default
      # # -- If true, enables commands execution from configured channel only.
      # restrictAccess: false

actions:
  'describe-created-resource':
    # -- If true, enables the action.
    enabled: false
    # -- Action display name posted in the channels bound to the same source bindings.
    displayName: "Describe created resource"
    # -- Command to execute when the action is triggered. You can use Go template (https://pkg.go.dev/text/template) together with all helper functions defined by Slim-Sprig library (https://go-task.github.io/slim-sprig).
    # You can use the `{{ .Event }}` variable, which contains the event object that triggered the action. See all available event properties on https://github.com/kubeshop/botkube/blob/main/pkg/event/event.go.
    # @default -- See the `values.yaml` file for the command in the Go template form.
    command: "kubectl describe {{ .Event.Kind | lower }}{{ if .Event.Namespace }} -n {{ .Event.Namespace }}{{ end }} {{ .Event.Name }}"

    # -- Bindings for a given action.
    bindings:
      # -- Event sources that trigger a given action.
      sources:
        #- plugin-based
      # -- Executors configuration used to execute a configured command.
      executors:
        - kubectl-plugin
plugins:
  repositories:
    botkube:
      url: http://localhost:8080/plugins-index.yaml

### To See Notifications in Action

```shell
kubectl run s9 --image=nginx:latest -n botkube
```

```shell
kubectl create configmap my-config-def --from-literal=key1=config1 -n botkube
```

and then update configmap to see diff is shown in slack

## Related issue(s)
Relates #840


<!-- If you refer to a particular issue, provide its number.
To close the issue after the pull request merge, use `Resolves #123` or `Fixes #123`.
Otherwise, use `See also #123` or just `#123`. -->

mszostok commented on pkg/api/executor/grpc_adapter.go:
```diff
@@ -53,7 +53,7 @@ type (
 		//	Old approach:
 		//	  return executor.ExecuteOutput{
-		//	  	Data: data,
+		//	  	Event: data,
```
please revert
```suggestion
		//	  	Data: data,
```

mszostok commented on pkg/api/source/grpc_adapter.go:
```diff
@@ -172,6 +218,22 @@ func (p *grpcServer) Stream(req *StreamRequest, gstream Source_StreamServer) err
 			}
+		case msg, ok := <-stream.Message:
+			if !ok {
+				return nil // output closed, no more chunk logs
```
```suggestion
				return nil // output closed, no more messages
```

mszostok commented on pkg/event/event.go:
```diff
@@ -71,7 +71,10 @@ var LevelMap = map[config.EventType]config.Level{
 func New(objectMeta metaV1.ObjectMeta, object interface{}, eventType config.EventType, resource string) (Event, error) {
 	typeMeta := k8sutil.GetObjectTypeMetaData(object)
 	event := Event{
-		TypeMeta:   typeMeta,
```
I know that we will fix that in the follow-up task. But let's remove that already so we will have a better diff with `internal/source/kubernetes/event/event.go` - in the rest of the code you can import that moved event instead.

huseyinbabal replied:
Isn't that dangerous to depend a public package to a plugin internals? I just wanted to wait until platform is removed, this event package is still used in a couple of places.

mszostok replied:
it's not, actually it's a proper solution to import some parts from `internal` in `pkg`. If you want I can explain on 1:1 call

mszostok commented on pkg/k8sutil/resource.go:
```diff
@@ -14,6 +14,11 @@ import (
 	"k8s.io/client-go/dynamic"
```
please remove that file and use `internal/source/kubernetes/k8sutil/resource.go` instead.

huseyinbabal replied:
As described above, I didn't wanted to depend our core to a plugin internal. I will be removing k8sutils anyways after platform migration, wdyt?

mszostok reviewed (commented):
I still need to check 5 more files:
- internal/source/dispatcher.go
- internal/source/kubernetes/registration.go
- internal/source/kubernetes/router.go
- internal/source/kubernetes/source.go
- internal/source/scheduler.go

And run some manual tests.

mszostok commented:
You need to move `filters` under `k8s` source:
https://github.com/kubeshop/botkube/blob/09e4a265cc182ee259638bf2daf13e1aac016c7a/helm/botkube/values.yaml#L411-L420

mszostok commented:
As a part of this PR, please also document all breaking under: https://github.com/kubeshop/botkube/issues/972. Thanks! 🙇

mszostok commented on internal/source/kubernetes/source.go:
```diff
@@ -0,0 +1,667 @@
+          "type": "array",
+          "items": {
+            "type": "string",
+            "enum": [
```
probably copy-past error :)

pkosiec commented on internal/source/kubernetes/config/config.go:
```diff
@@ -0,0 +1,307 @@
+
+// KubernetesFilters contains configuration for Kubernetes-related filters.
+type KubernetesFilters struct {
+	// ObjectAnnotationChecker enables support for `botkube.io/disable` and `botkube.io/channel` resource annotations.
```
Yes, I think we can remove `botkube.io/disable`, but I know at least a few people use it, so we should introduce `exclude` annotations and labels instead as a follow-up 🤔

huseyinbabal replied:
I completely moved the logic, and didn't wanted to touch internals. Handling as followup is fine by me.

mszostok replied:
OK, to do that in the follow-up 👍

pkosiec commented on cmd/executor/gh/main.go:
```diff
@@ -105,15 +105,15 @@ func (*GHExecutor) Help(context.Context) (api.Message, error) {
 
 var depsDownloadLinks = map[string]api.Dependency{
-	// Links source: https://github.com/cli/cli/releases/tag/v2.21.2
+	// Links source: https://github.com/cli/cli/releases/tag/v2.22.1
```
Actually it is upgrading (but there's 1.23.0 already)

huseyinbabal replied:
I will revert and test it again, I had a problem with gh plugin generation at that time, I will test with up to date version

pkosiec reviewed (commented):
Great work! Looks like the basic functionality works OK, I tested a few configurations and they worked well. Telemetry also OK 👍 

I agree with all @mszostok comments, and here are just a few ones from my side 🙂

huseyinbabal commented on internal/source/kubernetes/k8sutil/resource.go:
```diff
@@ -0,0 +1,103 @@
+package k8sutil
```
Let me explain;

As you already know, we haven't implemented all the platforms yet, platform renderers depends on `event` package, and `event` depends on k8sutil, that's why I keep them, we can remove that after platform migration, wdyt?

huseyinbabal commented on internal/source/kubernetes/source.go:
```diff
@@ -0,0 +1,667 @@
+func jsonSchema() api.JSONSchema {
+	return api.JSONSchema{
+		Value: heredoc.Docf(`
+{
```
I tested other plugins' json specs and it gives schema not found on validation, can we elaborate on that part?

huseyinbabal commented on internal/source/kubernetes/source.go:
```diff
@@ -0,0 +1,667 @@
+          "type": "array",
+          "items": {
+            "type": "string",
+            "enum": [
```
showing possible values will prevent flakiness, and I thought that, time to time we can change that options wdyt? I put current values we have in values.yaml, but I can definitely extend that to all possible values

huseyinbabal commented on internal/source/kubernetes/config/config.go:
```diff
@@ -0,0 +1,307 @@
+// Config Kubernetes configuration
+type Config struct {
+	KubeConfig           string `yaml:"kubeConfig"`
+	ClusterName          string
```
ok, I will remove them from config and pass the existing values from our configuration.

huseyinbabal commented on internal/source/kubernetes/event/event.go:
```diff
@@ -0,0 +1,145 @@
+	Error           string
+	Level           config.Level
+	Cluster         string
+	Channel         string
```
There is one usage in node annotation checker, is it ok to keep it for now?

huseyinbabal commented on cmd/source/kubernetes/main.go:
```diff
@@ -0,0 +1,19 @@
+package main
```
I just saw this, never happened to me but this is something we need to spot if we see that second time.

huseyinbabal commented on cmd/source/kubernetes/main.go:
```diff
@@ -0,0 +1,19 @@
+package main
+
```
That is because slack renderer has very special functionality to render kubernetes metadata like that, so rendering used to be handled on slack_renderer. However, in plugin structure, we render messages in advance, and I tried to unified them as a general message format. Again, I am open to suggestions here :)

huseyinbabal replied:
followup

huseyinbabal replied:
@mszostok is on it

huseyinbabal commented on internal/source/kubernetes/config/config.go:
```diff
@@ -0,0 +1,307 @@
+	Recommendations      *Recommendations   `yaml:"recommendations"`
+	Event                *KubernetesEvent   `yaml:"event"`
+	Resources            []Resource         `yaml:"resources" validate:"dive"`
+	ActionVerbs          []string           `yaml:"actionVerbs" validate:"dive"`
```
@pkosiec we don't enable plugins if they are not bound to channels. @mszostok if you are ok too, I will change that logic in a way that plugins will be always enabled.

huseyinbabal replied:
Basically, we have command in the context of actions only, that is why I have added like that.

huseyinbabal replied:
changed like suggested

huseyinbabal commented on cmd/source/kubernetes/main.go:
```diff
@@ -0,0 +1,19 @@
+package main
+
+import (
```
@pkosiec we don't enable plugins if they are not bound to channels. @mszostok if you are ok too, I will change that logic in a way that plugins will be always enabled.

huseyinbabal replied:
followup

huseyinbabal replied:
https://github.com/kubeshop/botkube/issues/1000

huseyinbabal commented on internal/source/kubernetes/commander/guard.go:
```diff
@@ -0,0 +1,258 @@
+package commander
```
followup task : extract to pkg to share between kubectl and k8s

huseyinbabal replied:
I handled that in this PR

mszostok commented on internal/source/scheduler.go:
```diff
@@ -97,27 +112,31 @@ func (d *Scheduler) schedule(ctx context.Context, bindSources []string) error {
 	// Holds the array of configs for a given plugin.
 	// For example, ['botkube/kubernetes@v1.0.0']->[]{"cfg1", "cfg2"}
```
Please update the comment to reflect current approach. Thanks!

mszostok commented on pkg/api/source/grpc_adapter.go:
```diff
@@ -23,13 +25,50 @@ type (
 	StreamInput struct {
 		// Configs is a list of Source configurations specified by users.
 		Configs []*Config
```
Right now there is no option that we will get a list of configs as we start multi-processes. When we will change do documentation and simplify this logic here?

Should we also do that for `executors`? 
Maybe we should do that as a follow-up after RBAC? WDYT @pkosiec @huseyinbabal?

huseyinbabal replied:
I can create a followup task for this, on my list. I also want to unify that, didn't wanted to confuse by having single config in sources and multiple configs in executors. Noted followup task.

huseyinbabal replied:
Followup https://github.com/kubeshop/botkube/issues/1001

mszostok commented on internal/source/kubernetes/source.go:
```diff
@@ -0,0 +1,687 @@
+}
+
+// Stream streams Kubernetes events
+func (*Source) Stream(ctx context.Context, input source.StreamInput) (source.StreamOutput, error) {
+	s := Source{}
+	s.startTime = time.Now()
+	s.messageCh = make(chan source.Message)
+	out := source.StreamOutput{Message: s.messageCh}
+	cfg, err := config.MergeConfigs(input.Configs)
+	if err != nil {
+		return source.StreamOutput{}, fmt.Errorf("while merging input configs: %w", err)
+	}
+	s.config = cfg
+	s.logger = loggerx.New(loggerx.Config{
+		Level: cfg.Log.Level,
+	})
+	s.ctx = ctx
+	s.clusterName = input.Context.ClusterName
+	s.kubeConfig = input.Context.KubeConfig
+	go consumeEvents(s)
+	return out, nil
+}
```
Let's clean-up the code in this function.
```go
func (*Source) Stream(ctx context.Context, input source.StreamInput) (source.StreamOutput, error) {
	cfg, err := config.MergeConfigs(input.Configs)
	if err != nil {
		return source.StreamOutput{}, fmt.Errorf("while merging input configs: %w", err)
	}

	s := Source{
		startTime: time.Now(),
		messageCh: make(chan source.Message),
		config:    cfg,
		logger: loggerx.New(loggerx.Config{
			Level: cfg.Log.Level,
		}),
		ctx:         ctx,
		clusterName: input.Context.ClusterName,
		kubeConfig:  input.Context.KubeConfig,
	}

	go consumeEvents(s)

	return source.StreamOutput{
		Message: s.messageCh,
	}, nil
}
```

mszostok reviewed (commented):
Overall good work 🚀

Once all comments will be resolved I will play with it locally as Paweł did.
Please also track all work that you would like to do in the follow-ups. Thanks!

mszostok commented on helm/botkube/values.yaml:
```diff
@@ -135,260 +135,278 @@ sources:
     displayName: "Kubernetes Info"
```
please update line 118 from `kubernetes` to `botkube/kubernetes`

mszostok commented on internal/source/kubernetes/source.go:
```diff
@@ -0,0 +1,687 @@
+	recRunner, recCfg := s.recommFactory.New(s.config)
+	err := recRunner.Do(s.ctx, &e)
+	if err != nil {
+		s.logger.Errorf("while running recommendations: %w", err)
```
```suggestion
		s.logger.Errorf("while running recommendations: %w", err)
		return
```

huseyinbabal replied:
good catch

mszostok commented on pkg/api/source/grpc_adapter.go:
```diff
@@ -23,13 +25,50 @@ type (
+		// You can construct a complex message.data or just use one of our helper functions:
+		//   - api.NewCodeBlockMessage("body", true)
+		//   - api.NewPlaintextMessage("body", true)
+		Message chan Message
+	}
+
+	Message struct {
+		Data      api.Message
+		Metadata  any
+		Telemetry map[string]interface{}
 	}
```
```go
	StreamOutput struct {
		Event chan Event
	}

	Event struct {
		Message          api.Message
		Metadata         any
		AnonymizedFields map[string]interface{}
	}
```

WDYT about such terminology? @huseyinbabal @pkosiec ?

huseyinbabal replied:
I am ok with the wording, changing accordingly

huseyinbabal commented:
LGTM with minor comments 🚀 

> Overall good work 🚀
> 
> Once all comments will be resolved I will play with it locally as Paweł did. Please also track all work that you would like to do in the follow-ups. Thanks!

I addressed all the feedbacks.

mszostok commented on helm/botkube/values.yaml:
```diff
@@ -135,260 +135,278 @@ sources:
+          types:
+            - create
+            - delete
+            - error
```
NOTE: after this merging this PR we won't merge configuration anymore. As a result, enabling both `k8s-err-events` and 'k8s-all-events' results in duplicated `error` messages. Probably will be good to mention that in the release notes/docs.

huseyinbabal replied:
Ok, I will be stating that in both release docs and main doc

mszostok commented on cmd/botkube/main.go:
```diff
@@ -129,7 +128,7 @@ func run(ctx context.Context) error {
 		return reportFatalError("while loading k8s config", err)
 	}
-	dynamicCli, discoveryCli, mapper, err := getK8sClients(kubeConfig)
+	_, discoveryCli, _, err := getK8sClients(kubeConfig)
```
it will be good to get rid of not used clients

mszostok commented on internal/source/kubernetes/source.go:
```diff
@@ -0,0 +1,692 @@
+	}, nil
+}
+
+func consumeEvents(s Source, ctx context.Context) {
```
context.Context always should be as a first parameter

mszostok commented on pkg/execute/factory.go:
```diff
@@ -7,19 +7,18 @@ import (
 	"github.com/slack-go/slack"
 
 	"github.com/kubeshop/botkube/internal/audit"
+	config2 "github.com/kubeshop/botkube/internal/command"
```
```suggestion
	guard "github.com/kubeshop/botkube/internal/command"
```

let's rename that to make it more readable

mszostok commented on pkg/filterengine/filterengine.go:
```diff
@@ -10,6 +10,11 @@ import (
 	"github.com/kubeshop/botkube/pkg/maputil"
```
actually this file can be deleted, and in the test we should use `filters.DisableAnnotation` from k8s source. I will handle that on my PR, don't worry about that.

mszostok reviewed (approved):
LGTM 👍 but before merge please address my last comments, and this one too: https://github.com/kubeshop/botkube/pull/985#discussion_r1121533684

Also please wait for @pkosiec feedback and approve. Thanks!

pkosiec reviewed (approved):
🚀 Awesome work!

huseyinbabal commented:
Addressed all the feedbacks, will check locally and focus on docs changes.
